LFS-352: Delete subjects, forms, treatments, ...

acrowthe opened the pull request:
Delete buttons added to all pages except individual subjects.
Implemented on all tables

Still todo:
- Add popup to confirm deletion
- Refresh tables
- Handle errors (ex. couldn't delete subject as subject had a form)
- Implement delete buttons on form and subject pages

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Add delete and permission buttons to user dashboard
- Implemented user dashboard delete, needs dynamic update

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Add buttons to every table and form
TODO:
- Refresh tables or page after delete
- Add deletion confirmation popup
- Add to individual subject page
- Handle error if deletion failed
    (ex, couldn't delete subject due to form)

sashaandjic commented:
It would look better if the Actions column is pinned to the right
![image](https://user-images.githubusercontent.com/23243244/89564028-67ae0b80-d7ea-11ea-9e05-be38edcd5e30.png)

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Make DeleteButton component and use that
- LiveTable modified to use action components
- Callback to reload table implemented
- Confirmation dialog added to DeleteButton,
    implemented on all LiveTables

sashaandjic commented:
Subject and tumors that were referenced in the forms cannot be deleted (correct). However, user is not presented with an error message ( e.g.  Subject cannot be removed because it was referenced in XYZ form).

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Add error and current page deletion handling

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -147,6 +151,14 @@ function Form (props) {
     setSelectorDialogOpen(false);
   }
 
+  let handleDelete = () => {
```
This logic is repeated 3 times. It should be moved to the DeleteButton component and enabled via a property.

acrowthe replied:
Done. A callback for reloading is now optional, and the `shouldGoBack` property on DeleteButtons will run this logic if true

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Move page back logic into delete button

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -0,0 +1,109 @@
+      </Dialog>
+      <Dialog open={open} onClose={closeDialog}>
+        <DialogTitle disableTypography>
+          <Typography variant="h6">Delete {entry["@name"]}</Typography>
+        </DialogTitle>
+        <DialogContent>
+            <Typography variant="body1">Are you sure you want to delete {entry["@name"]}?</Typography>
```
![image](https://user-images.githubusercontent.com/651980/89674726-9f828500-d8b6-11ea-8cf7-466a7993d032.png)


The Identifier displayed in the delete confirmation dialog is very technical and meaningless to the user. Ideally it should match whatever is displayed in the title (e.g. "Delete 123: Dempgraphics Data ..." per screenshot above). Something generic works too, for example:

```
Delete confirmation

Are you sure you want to delete this form?

                         [Delete] [Cancel]
```

(where "form" becomes "subject", "questionnaire", etc depending on the entity being deleted).

acrowthe replied:
![image](https://user-images.githubusercontent.com/20056751/89680277-a57d6380-d8c0-11ea-8ac0-691d57b04f01.png)

Deletion messages have been updated to match either the first column of a live table or title of the page

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -0,0 +1,109 @@
+        </DialogContent>
+        <DialogActions className={classes.dialogActions}>
+            <Button variant="contained" color="secondary" size="small" onClick={() => handleDelete()}>Delete</Button>
+            <Button variant="contained" size="small" onClick={closeDialog}>Close</Button>
```
"Cancel" seems more fitting to me than "Close".

acrowthe replied:
![image](https://user-images.githubusercontent.com/20056751/89675257-a1007d00-d8b7-11ea-9aeb-6a30c444d6dc.png)
Close was copied from the user deletion dialog box: Would you like that field changed too, or just this button?

marta- replied:
Who knows where else we would need to change, there are probably other confirmation dialogs... Okay, let's keep it as "Close" in your confirmation dialog for now for consistency, and we'll do a cleanup round at a later time. Changing the user deletion confirmation dialog would be out of scope for this PR. Thanks!

sashaandjic replied:
> Who knows where else we would need to change, there are probably other confirmation dialogs... Okay, let's keep it as "Close" in your confirmation dialog for now for consistency, and we'll do a cleanup round at a later time. Changing the user deletion confirmation dialog would be out of scope for this PR. Thanks!

I will log this as a separate task.

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Adjust deletion messages

sashaandjic commented:
I am unable to delete subject that is
-  no longer referenced in a form
- never being associated with the form 
![image](https://user-images.githubusercontent.com/23243244/89684606-c8137a80-d8c8-11ea-9c5a-f60e08d8a558.png)

Perhapst we should replace dialog title "Error" with "Warning"?

sashaandjic commented:
When deleting forms from Forms or Dashboard, I still get user "unfriendly" confirmation dialog
![image](https://user-images.githubusercontent.com/23243244/89684885-48d27680-d8c9-11ea-9ec4-763cb732244e.png)

acrowthe commented:
@sashaandjic 

> I am unable to delete any subject regardless of if it is
> 
> * no longer referenced in a form
> * never being associated with the form

Have any reproduction steps? I have not encountered that in my testing, but that should be fixed

> When deleting forms from Forms or Dashboard, I still get user "unfriendly" confirmation dialog

Not sure what the best way would be to fix that. Currently, all table delete buttons are pulling the value of the first column as the name. That works nicely for subjects, but the value forms use there is not user friendly.

I could make some way to just state "Delete form" while leaving subjects as is, but then the user would lose the ability to verify the right form is being deleted

sashaandjic commented:
> Have any reproduction steps? I have not encountered that in my testing, but that should be fixed
Sorry, I was wrong... Had to make a new clone...

sashaandjic commented:
Ok, I figured it out what happened originally... 
- create subject A with demohraphic form
- create tumor T (associated with subject) A with tumor form
- you cannot delete neither A nor T as they are both referenced in coresponding forms
- delete both forms
- you will not be able to delete subject A because it is referenced in tumor T, however there is no way in UI to see this unless you open tumor T 
- the remedy fir this situation would be to be more specific in the warning message e.g. "This subject is referenced in x forms and tumor a, b, c...)

acrowthe commented:
That makes sense.

Don't think the server returns that information alongside the error response, which would make more specific error messages difficult to do accurately. I'll look into it on Monday more to confirm though.

acrowthe commented:
@sashaandjic 
> * the remedy for this situation would be to be more specific in the warning message e.g. "This subject is referenced in x forms and in tumors: a, b, c...)

I looked into what it would take to get a more detailed error message. Unfortunately, the server is just responding with the error message "Unable to commit changes to session."

There would be two ways I can see to implement more detailed messages:
1. Using the front end, query for a) all forms owned by the to be deleted subject and b) all subjects are children of the to be deleted subject.
2. Modify the back end to return an error message which contains this data

For 2, I couldn't find where delete logic is handled, so I'm not sure what that would involve. Do you know if 2 would be more viable than 1, or at least where I should look to determine that?

acrowthe committed:
LFS-352 Delete subjects, forms, treatments, ...
Add recursive delete support and improve error text and appearance
- Match error format from vocabulary
- Update form delete name to match form title
- Add error message from server response, if any
- Add warning color to buttons per LFS-520

sashaandjic commented:
> I looked into what it would take to get a more detailed error message. Unfortunately, the server is just responding with the error message "Unable to commit changes to session."
> 
> There would be two ways I can see to implement more detailed messages:
> 
> 1. Using the front end, query for a) all forms owned by the to be deleted subject and b) all subjects which are children of the to be deleted subject.
> 2. Modify the back end to return an error message which contains this data
> 
> For 2, I couldn't find where delete logic is handled, so I'm not sure what that would involve. Do you know if 2 would be more viable than 1, or at least where I should look to determine that?

^ @marta-  @OrdiNeu  Would you please help?

IntegralProgrammer commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -173,13 +174,19 @@ function Form (props) {
             {": " + (data?.questionnaire?.title || id || "")}
+            <DeleteButton
+              entryPath={data ? data["@path"] : "/Forms/"+id}
+              entryName={data?.subject?.identifier + ": " + (data?.questionnaire?.title || id || "")}
```
If `data?.subject?.identifier` returns `null` then `entryName` will begin with `null: `. Wouldn't a pattern like the one that you have used on *Subject.jsx:148* work better?

acrowthe replied:
Fixed. Changed to `data?.subject?.identifier || "Subject"`

acrowthe commented:
This branch has been updated to utilize the improved deletion flow from #262. Branch [LFS-352-test](https://github.com/ccmbioinfo/lfs/tree/LFS-352-test) has been created for testing.

Changes:
- Adjusted commit messages to match name convention
- Delete buttons appear yellow when the entry is referenced (per LFS-520)
- Updated displayed form name on table entries to match title of form
- Added ability to recursively delete items if normal deletion fails due to references
- Update error style on complete deletion fail for consistency with other error dialogs

sashaandjic commented:
Deletion works well. All subjects that are referenced in at least one form have an orange trash can. 
![image](https://user-images.githubusercontent.com/23243244/90475089-c6318e80-e0f4-11ea-81d6-3f812b8fede9.png)

The proper warning is displayed asking user if they want to delete all forms where the subject is referenced. This also includes any tumor and corresponding tumor forms (not sure if we should also specify this in the warning msg).
![image](https://user-images.githubusercontent.com/23243244/90475067-b9ad3600-e0f4-11ea-87ae-5f204fe92ea3.png)

The only problem that I ran into was building LFS-352-test branch. The build was failing and I had to use -Pquick option to test it.
![image](https://user-images.githubusercontent.com/23243244/90475030-a9955680-e0f4-11ea-96cf-c1678c44c3db.png)

acrowthe commented:
> The only problem that I ran into was building LFS-352-test branch. The build was failing and I had to use -Pquick option to test it.

I've looked into reproducing this build issue, but have not been able to. The step that failed downloads and installs 3rd party modules, so I suspect a network issue. Without seeing the initial error I cannot confirm though

sashaandjic reviewed (approved):
tested

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Adjust deletion URL to match LFS-521 changes

sashaandjic commented:
ready for re-test?

sashaandjic commented:
One very small suggestion:
![image](https://user-images.githubusercontent.com/23243244/92051847-16e2f180-ed5b-11ea-94d4-de4da495ce14.png)

The message is a bit confusing. First we say we cannot delete, but then we offer to delete everything. I would remove highlighted text and replace underlined "items" forms and subjects so the text from the screenshot goes like this:
Patient P1 is referenced in 11 forms and 2 subjects( T2 and T1).

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Adjust recursive deletion message

sashaandjic reviewed (approved):
tested

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -29,6 +29,7 @@ import {
 } from "@material-ui/core";
 
 import LiveTable from "../dataHomepage/LiveTable.jsx";
+import DeleteButton from "../dataHomepage/DeleteButton.jsx";
```
Out of scope for this PR, but isn't the `commons` module a better place for the livetable and the buttons? @sdumitriu @OrdiNeu

sdumitriu replied:
Yes it is, since we use the livetable in more than one context now.

Currently we use two kinds of tables, our own `LiveTable`, and `MaterialTable`, could we replace our own with `MaterialTable`, or is there some custom functionality that can't be done with the library one?

sdumitriu replied:
But moving it is out of scope for this PR.

acrowthe replied:
`Livetable` automatically queries the backend for data, either through a passed url or through a url inferred from the current path. Also only fetches data for the page you are looking at.
`MaterialTable` need to be provided with the data, or a function to retrieve said data.

marta- reviewed (approved):
The code looks right, didn't test

acrowthe commented:
Resolved minor css merge conflict

sashaandjic commented:
re-tested

IntegralProgrammer commented:
There is an issue that I believe is in the scope of this PR which needs to be addressed. When LFS-277 is merged in (it's already merged into `dev`) the *Delete* button is not present on the *Incomplete Forms* Card on the User dashboard.

marta- commented:
> There is an issue that I believe is in the scope of this PR which needs to be addressed. When LFS-277 is merged in (it's already merged into `dev`) the _Delete_ button is not present on the _Incomplete Forms_ Card on the User dashboard.

Agreed, It's in scope

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Add delete button to Incomplete Forms table implemented in LFS-277

acrowthe commented:
Delete button added to the `Incomplete Forms` card

IntegralProgrammer reviewed (approved):
Tested and looks good! Approved :heavy_check_mark:

sashaandjic commented:
The lock icon is only shown on the Dashboard. Is this by design?
![image](https://user-images.githubusercontent.com/23243244/93640841-3ddf2b80-f9c9-11ea-8970-d601c52eb9b3.png)
![image](https://user-images.githubusercontent.com/23243244/93640874-4df70b00-f9c9-11ea-98cc-0f1adcdc4ee8.png)

acrowthe commented:
The lock icons from the dashboard are there as a placeholder for future implementation. They are placed there to match the layout from the Jira images

sashaandjic commented:
I think this PR is good to go. I mean, merge into dev.

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Adjust deletion for updated subject page
- Add delete button to form card
- Adjust top subject delete button around new form button

sashaandjic commented:
re-tested; all good
![image](https://user-images.githubusercontent.com/23243244/94316119-c756a680-ff51-11ea-9450-8266beee6b99.png)

sashaandjic commented:
2 problems detected:
![image](https://user-images.githubusercontent.com/23243244/94505159-70163780-01d8-11eb-83e4-c49cfd46e4a0.png)
- the trash can circled in yellow is too small (should be the same size as "+" button, should be after the "+" button as well)
- the trash can icon in the right bottom corner belongs to no-one
- the trash can circled in blue is floating

![image](https://user-images.githubusercontent.com/23243244/94505296-c4211c00-01d8-11eb-8970-f6dae6d25a7c.png)
This trash can should be bigger (size of the "+" button) and aligned with the logout icon

sdumitriu commented:
I think the blu circled icon deletes the "Tumors" form, and the arrowed icon deletes the whole Tumor subject (as a child subject of the patient).

The size comments seem valid concerns though.

sdumitriu reviewed (changes requested):
- [x] Shown after deleting a subject from the subjects table:

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in DeleteButton (created by Context.Consumer)
```

- [x] Open two tabs of the Subjects table, delete a subject in one tab, delete the same subject in the other tab -> the 404 error is not handled, and the UI keeps on displaying the "Are you sure" dialog

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Fix minor bugs
- Add handling for 404 error when deleting non-existant items
- Fix console error from updating delete button in removed table entry

sashaandjic commented:
re-tested; the problem with multiple tab being open and deleting the same subject, has been fixed
![image](https://user-images.githubusercontent.com/23243244/94848060-703d4f80-03f1-11eb-8b38-ec26bddc4d12.png)

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  let handleDelete = () => {
+    let request_data = new FormData();
```
This seems not used, can it be deleted?

acrowthe replied:
Deleted and tested.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -17,6 +17,8 @@
 //  under the License.
 //
 
+import { warningColor } from "../../../../../../commons/src/main/frontend/src/themeStyle";
```
Nope, since all jsx sources end up in the aggregated-frontend module, paths should be relative to the same frontend directory.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -205,15 +210,15 @@ const questionnaireStyle = theme => ({
         opacity: 1,
         zIndex: "1010"
     },
-    formProvider: {
```
Isn't this needed anymore?

acrowthe replied:
Added back: Used in one place

acrowthe committed:
LFS-352: Delete subjects, forms, treatments, ...
Adjust code review comments
- Remove unused import and variable
- Readd incorrectly removed style

acrowthe commented:
Rebased to pull in split chunk fix for testing

sashaandjic commented:
2 small ui problems detected:
- it would be nice if the trash can next to the "+" button is the same size as "+" button
- when I delete the form for subject 1050 in the screenshot below
![image](https://user-images.githubusercontent.com/23243244/95117947-60867980-0717-11eb-97a5-00ca85425fdc.png)
the screen does not refreshes so I get this again: 
![image](https://user-images.githubusercontent.com/23243244/95117967-6b410e80-0717-11eb-9cae-2547ebd161cc.png)
insted of this:
![image](https://user-images.githubusercontent.com/23243244/95118009-827ffc00-0717-11eb-8710-073f8cde6125.png)

sashaandjic commented:
Another problem: I am unable to delete tumor in this situation
![image](https://user-images.githubusercontent.com/23243244/95118339-249fe400-0718-11eb-8522-725385ffaf34.png)

acrowthe commented:
Both UI issues will be handled in separate issues
- it would be nice if the trash can next to the "+" button is the same size as "+" button
- when I delete the form for subject 1050 in the screenshot below the screen does not refresh

sashaandjic commented:
> Both UI issues will be handled in separate issues
> 
> * it would be nice if the trash can next to the "+" button is the same size as "+" button
> * when I delete the form for subject 1050 in the screenshot below the screen does not refresh

https://phenotips.atlassian.net/browse/LFS-594
https://phenotips.atlassian.net/browse/LFS-595
